### PR TITLE
OGRID+POINTCLOUD: New package to use sonar data and publish pointcloud

### PR DESCRIPTION
**Pull request by RustyBamboo**

### CODE REVIEW

File: perception/sub8_batcave_ogrid/include/OGridGen.hpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/include/OGridGen.hpp]: You should probably privatize some of these variables

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: This is unknown isn't it? The depth changes

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: You should create this before making the timer

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: Same as above

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: You know how many bounds there are, so you know the size this vector needs to be. Itd be best to preallocate that.

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: You know this size too, `std::vector<int8_t> data(size);`

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: Ditto

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: Might want to log an error here

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (mattlangford) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: Does this produce a flipped ogrid? I remember doing this in numpy would always result in a flipped/upside down ogrid from what I was expecting

**Review by mattlangford — CHANGES_REQUESTED** [review]
Small changes to improve efficiency with vector preallocation.

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: Can you change any console output that says sonar to imaging sonar?
This will get confusing since we have 2 passive and imaging.

File: perception/sub8_batcave_ogrid/include/OGridGen.hpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/include/OGridGen.hpp]: Private class members should end with an underscore.

File: perception/sub8_batcave_ogrid/package.xml
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/package.xml]: Lol I don't think the package should be called batcave. Feel free to leave batcave mentions but not the name of the package or files.

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: I would recommend using lambda's instead of bind here, it's a lot more flexible and readable. (Just something to keep in mind)

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: What does bound hold mean here?

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: If you need to use the iterator variable because you're iterating through multiple collections, you should just do a normal for loop. `for(int i = 0; i < bound_hold.response.bounds.size(); i++)`...

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: I think someting like this would be cleaner (You don't wan't your c++ to look too much like old c):
> auto* polygons = &bounds;
> auto npts = bounds.size();
> cv::polylines(matOgrid, &polygons, &npts, 1, true, 255, 3, CV_8U);

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: _Ahem_, variables should be under_scored, not camelCased. Of course you could just use [this](https://github.com/davetcoleman/roscpp_code_format) and get rid of all my annoying style comments.

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: You could minimize the use of parenthesis here and elsewhere: `int(ogrid_size / resolution)`

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_batcave_ogrid/src/OGridGen.cpp]: Consider turning this into:
> auto out_it = data.begin();
> for(auto it = matOgrid.begin<uchar>(), int k = 0; it != matOgrid.end<uchar>(); ++it)
> {
>   out_it = *it;
>   ++out_it;
> }

It's less error prone and more efficient (let me know if I messed anything up, I didn't actually compile this.)

**Review by DSsoto — CHANGES_REQUESTED** [review]
Can you change the tf stuff to use tf2? All new code should use it.
I assume you tested this and it works.

I think you use casts way too much lol you should feel a little dirty when you do that, they are rarely necessary.

After test day and the bounds stuff gets resolved we can think about merging.

File: perception/sub8_batcave_ogrid/src/OGridGen.cpp
Comment (RustyBamboo) [reply perception/sub8_batcave_ogrid/src/OGridGen.cpp]: polyline wants an array of arrays...

**Comment by kingkevlar** [comment]
Ahem... _Langforded_

File: perception/sub8_pointcloud/CMakeLists.txt
Comment (kev-the-dev) [inline perception/sub8_pointcloud/CMakeLists.txt]: Pretty sure this specifies the libraries you create in this package, so should be ```batcave_ogrid_lib``` or w/e

File: perception/sub8_pointcloud/CMakeLists.txt
Comment (kev-the-dev) [inline perception/sub8_pointcloud/CMakeLists.txt]: Respectable meme, but should change to something more descriptive before merged.

File: perception/sub8_pointcloud/src/OGridGen.cpp
Comment (kev-the-dev) [inline perception/sub8_pointcloud/src/OGridGen.cpp]: should probably be in relative namespace (not start with /)

File: perception/sub8_pointcloud/src/OGridGen.cpp
Comment (kev-the-dev) [inline perception/sub8_pointcloud/src/OGridGen.cpp]: because this might get big, it would make sense to have this as a class variable that you update. You don't want to be creating another huge vector every time this runs.

File: perception/sub8_pointcloud/src/OGridGen.cpp
Comment (kev-the-dev) [inline perception/sub8_pointcloud/src/OGridGen.cpp]: There's probably a cleaner way to do this part of copying the opencv Mat to the ogrid bytes. Checkout [Mat.ptr](http://docs.opencv.org/2.4/modules/core/doc/basic_structures.html?highlight=clone#mat-ptr) which should just be a pointer to the underlying array of uint8

File: perception/sub8_pointcloud/src/OGridGen.cpp
Comment (kev-the-dev) [inline perception/sub8_pointcloud/src/OGridGen.cpp]: :)

**Review by kev-the-dev — APPROVED** [review]
Promising start for using imaging sonar data. Nothing obviously broken so fine to merge. MIght be nice to squash some of the latest commits that don't follow naming convention.

**Comment by sentree** [comment]
When #219 was merged, this pull request was rebased onto master and automatically built by the CI server. There were errors that I suggest you check out over on that site.

**Comment by RustyBamboo** [comment]
Project 'sub8_pointcloud' tried to find library 'blueview_tools'. How to fix...?

**Comment by DSsoto** [comment]
@RustyBamboo it's called `mil_blueview_tools`.

**Comment by DSsoto** [comment]
I'm ready to merge this when it builds. I think if you change [this line](https://github.com/uf-mil/mil_common/blob/master/drivers/mil_blueview_driver/CMakeLists.txt#L30) to mil_blueview_tools it should build.

 Also can you interactive rebase (`git rebase -i HEAD~<num_of_commits>`) your commits to change the message on those that don't follow [our commit message convention](https://github.com/uf-mil/mil_common/wiki/Code-Policy#commit-messages0)?

**Comment by kev-the-dev** [comment]
What happened with the history here? Don't think you rebased right.

**Comment by RustyBamboo** [comment]
I gave up.

**Comment by kev-the-dev** [comment]
Should build now that uf-mil/mil_common#38 is merged. Just got push something for it to rebuild (maybe comments?)

File: perception/sub8_pointcloud/README.md
Comment (sentree) [inline perception/sub8_pointcloud/README.md]: Why?

File: perception/sub8_pointcloud/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_pointcloud/src/OGridGen.cpp]: This is a lot of unnecessary whitespace

File: perception/sub8_pointcloud/src/Classification.cpp
Comment (DSsoto) [inline perception/sub8_pointcloud/src/Classification.cpp]: This shoud be ros logging

File: perception/sub8_pointcloud/src/Classification.cpp
Comment (DSsoto) [inline perception/sub8_pointcloud/src/Classification.cpp]: Dead code, leave a TODO comment if you want to remember to do stuff in the future

File: perception/sub8_pointcloud/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_pointcloud/src/OGridGen.cpp]: Dead code, leave TODO instead

File: perception/sub8_pointcloud/src/OGridGen.cpp
Comment (DSsoto) [inline perception/sub8_pointcloud/src/OGridGen.cpp]: Are you sure Point() will always be initialized to (0, 0), I think it's best to be explicit here.

**Review by DSsoto — COMMENTED** [review]
Approving pending comment responses

**Review by DSsoto — APPROVED** [review]

**Comment by DSsoto** [comment]
Looks good!
Sorry we held this up for so long, but it should be better from here on out.

**Review by mattlangford — APPROVED** [review]

**Pull request merged**
